File: app/schemas/campaign.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.campaign import CampaignTargetGroup
from app.schemas.text import normalize_unicode_text


class CampaignBase(BaseModel):
    title: str = Field(..., min_length=2, max_length=255)
    description: str = Field(..., min_length=2)
    image_url: str | None = Field(default=None, max_length=2048)
    target_group: CampaignTargetGroup
    start_date: datetime | None = None
    end_date: datetime | None = None
    is_active: bool = True

    _normalize_text = field_validator("title", "description", mode="before")(normalize_unicode_text)


class CampaignCreate(CampaignBase):
    pass


class CampaignUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=255)
    description: str | None = Field(default=None, min_length=2)
    image_url: str | None = Field(default=None, max_length=2048)
    target_group: CampaignTargetGroup | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    is_active: bool | None = None

    _normalize_text = field_validator("title", "description", mode="before")(normalize_unicode_text)


class CampaignRead(CampaignBase):
    id: int
    created_by_user_id: int | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: mobile/lib/models/campaign.dart
class Campaign {
  const Campaign({
    required this.id,
    required this.title,
    required this.description,
    required this.targetGroup,
    required this.createdAt,
    this.imageUrl,
    this.startDate,
    this.endDate,
    this.isActive = true,
    this.createdByUserId,
    this.updatedAt,
  });

  final int id;
  final String title;
  final String description;
  final String? imageUrl;
  final String targetGroup;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isActive;
  final int? createdByUserId;
  final DateTime createdAt;
  final DateTime? updatedAt;

  factory Campaign.fromJson(Map<String, dynamic> json) {
    return Campaign(
      id: json['id'] as int,
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      imageUrl: json['image_url'] as String?,
      targetGroup: json['target_group'] as String? ?? 'all_clients',
      startDate: _parseDate(json['start_date']),
      endDate: _parseDate(json['end_date']),
      isActive: json['is_active'] as bool? ?? true,
      createdByUserId: json['created_by_user_id'] as int?,
      createdAt: DateTime.tryParse(json['created_at']?.toString() ?? '') ??
          DateTime.now(),
      updatedAt: _parseDate(json['updated_at']),
    );
  }

  static DateTime? _parseDate(Object? value) {
    if (value == null) {
      return null;
    }
    return DateTime.tryParse(value.toString());
  }
}
